RecoCTPPS/TotemRPLocal/src/FastLineRecognition.cc

```diff
@@ -0,0 +1,342 @@
+
+FastLineRecognition::GeomData FastLineRecognition::GetGeomData(unsigned int id)
+{
+  // result already buffered?
+  map<unsigned int, GeomData>::iterator it = geometryMap.find(id);
+  if (it != geometryMap.end())
+    return it->second;
+
+  // calculate it
+  CLHEP::Hep3Vector d = geometry->LocalToGlobalDirection(id, CLHEP::Hep3Vector(0., 1., 0.));
+  DDTranslation c = geometry->GetDetector(id)->translation();
+  GeomData gd;
+  gd.z = c.z();
+  gd.s = d.x()*c.x() + d.y()*c.y();
+
+  geometryMap[id] = gd;
+
+  return gd;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void FastLineRecognition::GetPatterns(const DetSetVector<TotemRPRecHit> &input, double z0,
+  double threshold, DetSet<TotemRPUVPattern> &patterns)
+{
+  // build collection of points in the global coordinate system
+  std::vector<Point> points;
+  for (auto &ds : input)
+  {
+    unsigned int detId = ds.detId();
+
+    for (auto &h : ds)
+    {
+      const TotemRPRecHit *hit = &h;
+      const GeomData &gd = GetGeomData(detId);
+  
+      double p = hit->getPosition() + gd.s;
+      double z = gd.z - z0;
+      double w = sigma0 / hit->getSigma();
+  
+      points.push_back(Point(detId, hit, p, z, w));
+    }
+  }
+
+#if DEBUG > 0
+  printf(">> FastLineRecognition::GetPatterns(z0 = %E)\n", z0);
+  printf(">>>>>>>>>>>>>>>>>>>\n");
+#endif
+
+  // reset output
+  patterns.clear();
+
+  Cluster c;
+  while (GetOneLine(points, threshold, c))
+  {
+    // convert cluster to pattern and save it
+    TotemRPUVPattern pattern;
+    pattern.setA(c.Saw/c.Sw);
+    pattern.setB(c.Sbw/c.Sw);
+    pattern.setW(c.weight);
+
+#if DEBUG > 0
+    printf("\tpoints of the selected cluster: %lu\n", c.contents.size());
+#endif
+
+    for (auto &pit : c.contents)
+    { 
+#if DEBUG > 0
+      printf("\t\t%.1f\n", pit->z);
+#endif
+      pattern.addHit(pit->detId, *(pit->hit));
+    }
+
+    patterns.push_back(pattern);
+
+#if DEBUG > 0
+    unsigned int u_points_b = 0;
+    for (vector<Point>::iterator dit = points.begin(); dit != points.end(); ++dit)
+      if (dit->usable)
+        u_points_b++;
+    printf("\tusable points before: %u\n", u_points_b);
+#endif
+
+    // remove points belonging to the recognized line
+    for (vector<const Point *>::iterator hit = c.contents.begin(); hit != c.contents.end(); ++hit)
+    {
+      for (vector<Point>::iterator dit = points.begin(); dit != points.end(); ++dit)
+      {
+        //printf("\t\t1: %.2f, %p vs. 2: %.2f, %p\n", (*hit)->z, (*hit)->hit, dit->z, dit->hit);
+        if ((*hit)->hit == dit->hit)
+        {
+          dit->usable = false;
+          //points.erase(dit);
+          break;
+        }
+      }
+    }
+
+#if DEBUG > 0
+    unsigned int u_points_a = 0;
+    for (vector<Point>::iterator dit = points.begin(); dit != points.end(); ++dit)
+      if (dit->usable)
+        u_points_a++;
+    printf("\tusable points after: %u\n", u_points_a);
+#endif
+  }
+
+#if DEBUG > 0
```

Comment:
maybe CTPPS_DEBUG?
so that accidentally someone else's DEBUG doesn't enable this